### DISCUSSION TITLE
Record audio to file in sink mode and pay back as a source

### DISCUSSION THREAD

**Discussion by jmsunseri** [pschatzmann/ESP32-A2DP#475] [Q&A] [answered]
is it possible to create a file from the stream on a bluetooth sink in a format that can be played back when in source mode?

**Comment by pschatzmann** [comment, marked as answer]
Sure, it's both the same PCM format (as documented in the [Readme](https://github.com/pschatzmann/ESP32-A2DP)) but you can't run the sink and source at the same time.

However, I would recommend to store them as WAV files so that you can play the files with any other audio player easily. For this I can recommend my [AudioTools](https://github.com/pschatzmann/arduino-audio-tools/wiki)

**Comment by jmsunseri** [comment]
How do I go from a `set_raw_stream_reader` function on a ESP32 in sink mode to a wav file?  The only think I have figured out right now was to print all the channel 1/2 values out to serial and then pipe that to a text file.  The result has been just awful.

**Comment by pschatzmann** [comment]
You would normally write to a file on a sd drive. 

You are crazy if you try to send this over Serial! Did you calculate the required Baud rate to send stereo 16 bit audio with a sampe rate of 44100 ?

You would need to downsample the signal first.

**Reply by jmsunseri** [reply]
I do not have and SD card attached to this ESP32 module, that would make a lot of sense though.  This might explain why my results have been so poor.  Assuming i add that on is there a good example in the wiki for taking the `set_raw_stream_reader(const uint8_t *data, uint32_t frameCount)` arguments and building a wav file?

**Comment by pschatzmann** [comment]
No, but just read the chaper about encoding that i povided above. Instead of calling copy, you would just call write with your data
